### ISSUE TITLE
dill fails to dump function in subprocess

### ISSUE THREAD

**Issue by voltane** [uqfoundation/dill#96]
Here is my example code to reproduce the problem:
<pre>
import zmq
import time
import dill
from multiprocessing import Process

port = "71903"

def worker(curr_namespace):
    context = zmq.Context()
    w_socket = context.socket(zmq.PAIR)
    w_socket.bind("tcp://127.0.0.1:%s" % port)
    ns = dill.loads(w_socket.recv())
    namespace = curr_namespace.copy()
    exec(ns, namespace, namespace)
    result = {}
    for key, value in namespace.items():
        if key in curr_namespace.keys():
            if value == curr_namespace[key]:
                continue
        result[key] = value
    w_socket.send(dill.dumps(result)) # Dumps throw error: ValueError: I/O operation on closed file
    w_socket.close()
    return
    
p = Process(target=worker, args=(locals(),))
p.start()

context = zmq.Context()
socket = context.socket(zmq.PAIR)
socket.connect("tcp://127.0.0.1:%s"%port)
ns = "class A(object):\n\tpass" # This works
# ns = "def a():\n pass" # This one not
socket.send(dill.dumps(ns))
p.join()
a = dill.loads(socket.recv())
print a
socket.close()
</pre>
As you can see, i'm trying to execute some code in subprocess and then send namespace changes back. Everything works fine until i try to define new function.
Used Python: 2.7.6 and 2.7.8
Dill version: 0.2.2
I wonder if this is my problem or bug in pickling process.

**Comment by matsjoyce** [comment]
The problem is that as part of a multiprocessing `Process` object (like `p`), there is an attribute `authkey` of type `AuthenticationString`, which has the code:

```py
class AuthenticationString(bytes):
    def __reduce__(self):
        from .context import get_spawning_popen
        if get_spawning_popen() is None:
            raise TypeError(
                'Pickling an AuthenticationString object is '
                'disallowed for security reasons'
                )
        return AuthenticationString, (bytes(self),)
```

If you remove `Process` and `p` from the namespace:

```py
def worker(curr_namespace):
    context = zmq.Context()
    w_socket = context.socket(zmq.PAIR)
    w_socket.bind("tcp://127.0.0.1:%s" % port)
    ns = dill.loads(w_socket.recv())
    namespace = curr_namespace.copy()
    del namespace["Process"]
    del namespace["p"]
    exec(ns, namespace, namespace)
    result = {}
    for key, value in namespace.items():
        if key in curr_namespace.keys():
            if value == curr_namespace[key]:
                continue
        result[key] = value
    w_socket.send(dill.dumps(result)) # Dumps throw error: ValueError: I/O operation on closed file
    w_socket.close()
    return
```

The change being:

```py
    del namespace["Process"]
    del namespace["p"]
```

Results in:

```
{'a': <function a at 0x7fb117a466a8>}
```

I'd be tempted to say its not a dill bug, more a multiprocessing feature.

**Comment by voltane** [comment]
@matsjoyce thanks for your answer.
I tried your solution, but it doesn't solve the problem and i probably know why.
As you can see here:
<pre>
for key, value in namespace.items():
        if key in curr_namespace.keys():
            if value == curr_namespace[key]:
                continue
        result[key] = value
</pre>
I'm adding only new and edited variables to result namespace (p and Process are the same so i dont move them).
In fact, when i remove this for-loop, another exception is raised (about AuthenticationString). So it is not my main problem - still getting I/O Exception.

**Comment by matsjoyce** [comment]
Ah, I was lazy, and only tested under python 3, so the bug with the "fixed" version must be python 2 specific. I get with trace on:

```py
D2: <dict object at 0x7f3776f3cb40>
F1: <function a at 0x7f3776eec050>
F2: <function _create_function at 0x7f3776f1af50>
Co: <code object a at 0x7f3776edd8b0, file "<string>", line 1>
F2: <function _unmarshal at 0x7f3776f1ade8>
D2: <dict object at 0x7f3776f31280>
F1: <function a at 0x7f3776eec050>
D2: <dict object at 0x7f3776f2bd70>
M2: <module 'dill' from '/home/matthew/GitHub/dill/dill/__init__.pyc'>
F2: <function _import_module at 0x7f3776f1d848>
M2: <module '__builtin__' (built-in)>
M2: <module 'time' from '/usr/lib/python2.7/lib-dynload/time.so'>
F1: <function worker at 0x7f3776edbde8>
Co: <code object worker at 0x7f3780724ab0, file "a.py", line 9>
D1: <dict object at 0x7f37807c0168>
D2: <dict object at 0x7f3776f407f8>
M2: <module 'zmq' from '/usr/lib/python2.7/site-packages/zmq/__init__.pyc'>
Process Process-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "a.py", line 24, in worker
    w_socket.send(dill.dumps(result)) # Dumps throw error: ValueError: I/O operation on closed file
  File "/home/matthew/GitHub/dill/dill/dill.py", line 194, in dumps
    dump(obj, file, protocol, byref, fmode)#, strictio)
  File "/home/matthew/GitHub/dill/dill/dill.py", line 184, in dump
    pik.dump(obj)
  File "/usr/lib/python2.7/pickle.py", line 224, in dump
    self.save(obj)
  File "/usr/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/home/matthew/GitHub/dill/dill/dill.py", line 678, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/lib/python2.7/pickle.py", line 649, in save_dict
    self._batch_setitems(obj.iteritems())
  File "/usr/lib/python2.7/pickle.py", line 686, in _batch_setitems
    save(v)
  File "/usr/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/home/matthew/GitHub/dill/dill/dill.py", line 646, in save_function
    obj.__dict__), obj=obj)
  File "/usr/lib/python2.7/pickle.py", line 405, in save_reduce
    self.memoize(obj)
  File "/usr/lib/python2.7/pickle.py", line 244, in memoize
    assert id(obj) not in self.memo
AssertionError
```

**Comment by matsjoyce** [comment]
OK, the problem is, one `p` and `Process` has been removed, is that when pickling the function `a`, pickle serialises the globals of `a`, and inside the globals of `a` is `a`. This means it stumbles upon its self, and pickle bales out. This doesn't happen in 3.4 due to https://github.com/python/cpython/blob/master/Lib/pickle.py#L603-L609 which for some reason was changed under PEP 3154. dill usually jumps in and helps out by changing the globals to a pointer to a modules globals, so there's no recursion, but since your evaling, the globals are not kept anywhere else.
I'm not quite sure how to fix this apart from moving to 3.4 yet, though, as its a kind of intrinsic pickle problem.

**Comment by mmckerns** [comment]
@matsjoyce: this might also be fixed by #86.  I have built a function off of the patch in #86 that allows one to set the depth into globals one would like to go, and pick whether you want a pointer to globals, all the globals in the global scope, or only the globals in the local scope.  I'm guessing it's also what's needed here… but I will test after the patch is ready to be applied.

**Comment by matsjoyce** [comment]
Yeah, it sounds like it will. It seems 99% of function problems are globals related.

**Comment by voltane** [comment]
Anyway, thanks for answering. 
Currently I switched to Python 3.x and now its work like a charm.

**Comment by mmckerns** [comment]
@matsjoyce: I just tested it, and this is not improved by the fix for #100.
